packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+    _animation.addListener(_onAnimate);
+    _controller.forward();
+  }
+
+  // Handle mousewheel scroll events.
+  void _receivedPointerSignal(PointerSignalEvent event) {
+    if (event is PointerScrollEvent) {
+      final Size childSize = _getChildSize(
+        _childKey.currentContext.findRenderObject() as RenderBox,
+        _constraints,
+      );
+      final double scaleChange = 1.0 + event.scrollDelta.dy / childSize.height;
+      if (scaleChange == 0.0) {
+        return;
+      }
+      final Offset focalPointScene = widget.transformationController.toScene(
+        event.localPosition,
+      );
+      setState(() {
+        widget.transformationController.value = _matrixScale(
+          widget.transformationController.value,
+          scaleChange,
+        );
+
+        // After scaling, translate such that the event's position is at the
+        // same scene point before and after the scale.
+        final Offset focalPointSceneScaled = widget.transformationController.toScene(
+          event.localPosition,
+        );
+        widget.transformationController.value = _matrixTranslate(
+          widget.transformationController.value,
+          focalPointSceneScaled - focalPointScene,
+        );
+      });
+    }
+  }
+
+  // Handle inertia drag animation.
+  void _onAnimate() {
+    if (!_controller.isAnimating) {
+      _animation?.removeListener(_onAnimate);
+      _animation = null;
+      _controller.reset();
+      return;
+    }
+    // Translate such that the resulting translation is _animation.value.
+    final Vector3 translationVector = widget.transformationController.value.getTranslation();
+    final Offset translation = Offset(translationVector.x, translationVector.y);
+    final Offset translationScene = widget.transformationController.toScene(
+      translation,
+    );
+    final Offset animationScene = widget.transformationController.toScene(
+      _animation.value,
+    );
+    final Offset translationChangeScene = animationScene - translationScene;
+    setState(() {
+      widget.transformationController.value = _matrixTranslate(
+        widget.transformationController.value,
+        translationChangeScene,
+      );
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    _controller = AnimationController(
+      vsync: this,
+    );
+  }
+
+  @override
+  void didUpdateWidget(InteractiveViewer oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.child != oldWidget.child
+      || widget.boundaryMargin != oldWidget.boundaryMargin) {
+      _boundaryRectCached = null;
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // A GestureDetector allows the detection of panning and zooming gestures on
+    // the child.
+    return LayoutBuilder(
+      builder: (BuildContext context, BoxConstraints constraints) {
+        _constraints = constraints;
```

Comment:
Caching the constraints here seems a little sketchy. When you need the constraints, can you not just get them from your own RenderObject (via `context.getRenderObject().constraints`)?

Reply:
It's also not super-clear to me why we need the constraints at all?

Reply:
This is still not clear to me. It appears its used to calculate the size of the "viewport". Can you just ask one of the parents around the child widget what its RenderBox size is to get that information?